src/platform_impl/linux/x11/event_processor.rs

```diff
@@ -669,710 +621,685 @@ impl<T: 'static> EventProcessor<T> {
-                            self.kb_state.update_modifiers(
-                                xev.base_mods,
-                                xev.latched_mods,
-                                xev.locked_mods,
-                                xev.base_group as u32,
-                                xev.latched_group as u32,
-                                xev.locked_group as u32,
-                            );
-                            let new_mods = self.kb_state.mods_state();
-                            if prev_mods != new_mods {
-                                if let Some(window) = self.active_window {
-                                    callback(Event::WindowEvent {
-                                        window_id: mkwid(window),
-                                        event: WindowEvent::ModifiersChanged(
-                                            Into::<ModifiersState>::into(new_mods).into(),
-                                        ),
-                                    });
-                                }
-                            }
-                        }
-                        _ => {}
+            X11Event::XkbStateNotify(xev) => {
+                // Set the timestamp.
+                wt.xconn.set_timestamp(xev.time as xproto::Timestamp);
+
+                let prev_mods = self.kb_state.mods_state();
+                self.kb_state.update_modifiers(
+                    xev.base_mods.into(),
+                    xev.latched_mods.into(),
+                    xev.locked_mods.into(),
+                    xev.base_group as u32,
+                    xev.latched_group as u32,
+                    xev.locked_group.into(),
+                );
+                let new_mods = self.kb_state.mods_state();
+                if prev_mods != new_mods {
+                    if let Some(window) = self.active_window {
+                        callback(Event::WindowEvent {
+                            window_id: mkwid(window),
+                            event: WindowEvent::ModifiersChanged(
+                                Into::<ModifiersState>::into(new_mods).into(),
+                            ),
+                        });
                     }
                 }
-                if event_type == self.randr_event_offset as c_int {
-                    self.process_dpi_change(&mut callback);
-                }
             }
-        }
 
-        // Handle IME requests.
-        if let Ok(request) = self.ime_receiver.try_recv() {
-            let mut ime = wt.ime.borrow_mut();
-            match request {
-                ImeRequest::Position(window_id, x, y) => {
-                    ime.send_xim_spot(window_id, x, y);
-                }
-                ImeRequest::Allow(window_id, allowed) => {
-                    ime.set_ime_allowed(window_id, allowed);
-                }
+            X11Event::RandrNotify(_) => {
+                self.process_dpi_change(&mut callback);
```

Comment:
I came here for another reason, but then noticed this bit. I think this is wrong.

The old code does:
```
if event_type == self.randr_event_offset as c_int {
```
Imagine a `+ 0` in there and check that "event zero" from randr is `RRScreenChangeNotify`: https://docs.rs/x11-dl/latest/x11_dl/xrandr/constant.RRScreenChangeNotify.html

RandrNotify is event 1: https://docs.rs/x11-dl/latest/x11_dl/xrandr/constant.RRNotify.html